### PR TITLE
add Graph

### PR DESCRIPTION
**Pull request by Gummibeer**
**solves:** #43 

```php
$graph = new Graph();

// create product and prelink organization
$graph->product()->name('My cool Product')->brand($graph->organization());
// hide the organization from the created script tag
$graph->hide(\Spatie\SchemaOrg\Organization::class);

// somewhere else fill organization
$graph->organization()->name('My awesome Company');

// render graph to script tag
echo $graph;
```

**output** (beautified)
```html
<script type="application/ld+json">
{
   "@context":"https:\/\/schema.org",
   "@graph":[
      {
         "@type":"Product",
         "name":"My cool Product",
         "brand":{
            "@type":"Organization",
            "name":"My awesome Company"
         }
      }
   ]
}
</script>
```

### CODE REVIEW

**Comment by Gummibeer** [comment]
Hey @sebastiandedeyne ,
can you please review this PR?

**Comment by sebastiandedeyne** [comment]
Just back from vacation and pretty scarce on time with client work, I'm going to go through this some time in the next 2-3 weeks. Sorry for the delay!

**Comment by Gummibeer** [comment]
@nolros Can you review this? For me it work's like a charm but I don't know if it solves your problem from #43 ?

**Comment by Gummibeer** [comment]
@spatie @spatie-bot @sebastiandedeyne - has anyone time to review this PR, and also the other open ones!?

**Comment by sebastiandedeyne** [comment]
Hi @Gummibeer,

First of all, sorry for the major delay!

I generally like this PR, but it kinda breaks with the idea that you can call everything in this package fluently.

How about something like this?

```php
echo Schema::graph()
    ->product(function (Product $product, Graph $graph) {
        return $product
            ->name('My cool Product')
            ->brand($graph->organization());
    })
    ->organization(function (Organization $organization) {
        return $organization->name('My awesome Company');
    })
    ->hide(Organization::class);
```

Since this has been so long, I'd totally understand if you're not interested in working on this PR anymore, and I'll make the necessary changes myself!

**Comment by Gummibeer** [comment]
@sebastiandedeyne thanks for your response! 😀 
So your issue is the return of the `getOrCreate()` method!? Or better that this method is used if you call `$graph->organization()`!?
I don't want to change/remove the `getOrCreate()` method because it's so usefull with the cache. 
But I like your idea to have two cases - `__call` called with a callback and without.
I will rework this! :)

**Comment by Gummibeer** [comment]
@sebastiandedeyne it's done! :)
I hope this is what you wanted!? It works like what you described.^^

File: src/Graph.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        if (!$this->has($type)) {
+            $this->set(new $type());
+        }
+
+        return $this->get($type);
+    }
+
+    public function hide(string $type)
```
Comment (peter279k) [inline src/Graph.php]: I think It should add the return type hint named `self`.

Comment (Gummibeer) [reply src/Graph.php]: I don't have good experience with `self` return types. If it comes to overriding you can't use `self` but have to use the extended classname. 😣
I'm open for the CR but want to share my thoughts why it hasn't happened. Even if I think that this class won't be extended/overriden this much.

Comment (peter279k) [reply src/Graph.php]: @Gummibeer, my confused problem is about why some methods misses the `return type hint`.

And I just think they should be added :).

Comment (Gummibeer) [reply src/Graph.php]: https://github.com/spatie/schema-org/pull/62/commits/452858e3b39ac06fba4ae1e5dfc6c94d87aa4082

File: generator/templates/static/Graph.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+namespace Spatie\SchemaOrg;
+
+use ReflectionClass;
+use BadMethodCallException;
+use InvalidArgumentException;
+
+/**
+ * @mixin \Spatie\SchemaOrg\Schema
+ */
+class Graph extends BaseType
```
Comment (sebastiandedeyne) [inline generator/templates/static/Graph.php]: I made some minor code style changes to this class. Could you rerun the generator and commit to update the file in `src/`?

File: generator/templates/static/Graph.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+        throw new BadMethodCallException(sprintf('The method "%" does not exist on class "%s".', $method, get_class($this)));
+    }
+
+    public function add(Type $schema): self
+    {
+        $type = get_class($schema);
+        
+        if ($this->has($type)) {
+            throw new InvalidArgumentException(sprintf('The graph already has an item of type "%s".', $type));
```
Comment (sebastiandedeyne) [inline generator/templates/static/Graph.php]: Could this be extracted to it's own exception? Maybe a `TypeAlreadyInGraph`?

File: generator/templates/static/Graph.php [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    public function set(Type $schema)
+    {
+        return $this->setProperty(get_class($schema), $schema);
+    }
+
+    public function get(string $type): Type
+    {
+        if (! $this->has($type)) {
+            throw new InvalidArgumentException(sprintf('The graph does not have an item of type "%s".', $type));
```
Comment (sebastiandedeyne) [inline generator/templates/static/Graph.php]: Same as above, but `TypeNotInGraph`

**Pull request merged**